test/unreliable_feed.js

```diff
@@ -0,0 +1,125 @@
+// Copyright © 2017 IBM Corp. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+const debug = require('debug')('cloudant-follow:test:feed_changes');
+const follow = require('../api');
+const spawn = require('child_process').spawn;
+const tap = require('tap');
+
+// mock server options:
+const port = 3000;
+const totalUpdates = 500;
+
+var mockServer;
+
+tap.beforeEach(function(done) {
+  mockServer = spawn('node', [
+    './test/mocks/mock_server.js',
+    // CLI arguments:
+    '--abort-every', 59,
+    '--delay-probability', 0.1,
+    '--error-probability', 0.3,
+    '--heartbeat', 6000,
+    '--last-seq-every', 181,
+    '--min-delay', 100,
+    '--max-delay', 5000,
+    '--port', port,
+    '--total-updates', totalUpdates
+  ]);
+  // capture stdout/stderr
+  mockServer.stderr.on('data', (data) => {
+    debug(data.toString());
+  });
+  mockServer.stdout.on('data', (data) => {
+    debug(data.toString());
+    if (data.toString().startsWith('Listening on port')) {
+      done();
+    }
+  });
+});
+
+tap.afterEach(function(done) {
+  mockServer.kill();
+  done();
+});
+
+tap.test('Captures all changes in unreliable `/_changes` feed', { timeout: 300000 }, function(t) {
+  var actualChanges = [];
+  var expectedChanges = [];
+
+  // build expected changes array
+  for (let i = 1; i <= totalUpdates; i++) {
+    expectedChanges.push(`doc${i}`);
+  }
+
+  var feed = new follow.Feed();
+  feed.db = `http://localhost:${port}/foo`;
+
+  feed
+    .on('error', function(error) {
+      t.fail(`Unexpected error: ${error}`);
+    })
+    .on('change', function(change) {
+      let i = actualChanges.indexOf(change.id);
+      t.equal(i, -1, `Unseen change for doc '${change.id}'.`, { bail: true });
+      t.ok(change.seq, 'Valid seq for change.');
+
+      actualChanges.push(change.id);
+      if (i !== -1 || actualChanges.length === totalUpdates) {
+        feed.stop();
+      }
+    })
+    .on('stop', function() {
+      t.deepEqual(actualChanges, expectedChanges);
```

Comment:
Are there guarantees about ordering of sequence numbers on the changes feed? I guess you don't have to worry about this because the test will only ever be used in conjunction with the mock server, but it might be relevant

Reply:
Yeah, the response from the mock is guaranteed to be ordered. The thing I guess I'm really testing is the right `since` values are being used on retry as to avoid seeing duplicate changes come through.